Update model notebook with filter parameter (#149)
AlexsLemonade/scpca-paper-figures: analysis/pseudobulk-bulk-prediction/model-notebooks/build-assess-models.Rmd

* `fit_additive`: `bulk ~ pseudobulk + (1|sample_id)`
* `fit_interaction`: `bulk ~ pseudobulk + (1|sample_id) + (1|sample_id:pseudobulk)`

> In addition, the notebook parameter `expr_threshold` determines how genes are filtered from the data before modeling: Only genes which are expressed in at least `2 * expr_threshold` in both modalities combined are retained.

Reviewer: The wording here is tough, but I think it is worth trying to describe the intent rather than the precise calculation.

But we definitely want to say "greater than" rather than "at least" because filtering at 0 does filter samples!
```suggestion
In addition, the notebook parameter `expr_threshold` determines how genes are filtered from the data before modeling: Only genes which are expressed in greater than the proportion `expr_threshold` of samples, averaged across bulk and pseudobulk modalities, are retained.
```

Author: thank you! boy was I having a fun time phrasing this, which is why I gave up and just showed the math

pseudobulk_files <- list.files(
  path = file.path(data_dir, "pseudobulk"),
  full.names = TRUE,
  pattern = "_pseudobulk\\.tsv$"
)
pseudobulk_names <- stringr::str_split_i(basename(pseudobulk_files), pattern = "_", i = 1)
names(pseudobulk_files) <- pseudobulk_names

bulk_counts_file <- file.path(data_dir, "normalized-bulk-counts.rds")
bulk_counts <- readr::read_rds(bulk_counts_file)

# Make sure we have the same projects, in the same order
stopifnot(
  all.equal(names(pseudobulk_files), names(bulk_counts))
)

# also the tables with fraction expressed
bulk_frac_expr_df <- readr::read_tsv(

>   file.path(data_dir, "fraction-expressed-bulk.tsv")

Reviewer: may as well skip this one too.
```suggestion
  file.path(data_dir, "fraction-expressed-bulk.tsv"),
  show_col_types = FALSE
```

) |>
  dplyr::rename(bulk_frac_expressed = frac_samples_expressed)
  
named_projects <- purrr::set_names(pseudobulk_names, pseudobulk_names)
expressed_df <- named_projects |>
  purrr::map(
    \(project_id) {
      readr::read_tsv(

>         file.path(data_dir, glue::glue("{project_id}_fraction-expressed-single-cell.tsv"))

Reviewer: Avoid repeated col_type printing.
```suggestion
        file.path(data_dir, glue::glue("{project_id}_fraction-expressed-single-cell.tsv")),
        show_col_types = FALSE
```

      )
    }
  ) |>
  purrr::list_rbind(names_to = "project_id") |>
  dplyr::rename(pseudobulk_frac_expressed = frac_samples_expressed) |>
  dplyr::full_join(bulk_frac_expr_df, by = c("project_id", "ensembl_id"))
```

### Functions

This chunk defines a helper function to read and combine bulk and pseudobulk datasets for a single project.
```{r}
prepare_data <- function(pseudo_file, bulk_df, id, exclude_samples, expressed_df) {
  
  # determine genes to keep
  keep_genes <- expressed_df |>
    dplyr::filter(
      project_id == id, 

>       (pseudobulk_frac_expressed + bulk_frac_expressed) > (params$expr_threshold * 2)

Reviewer: Since I was saying something about the intent, we might as well put that in the code more directly.
```suggestion
      mean(pseudobulk_frac_expressed, bulk_frac_expressed) > params$expr_threshold
```

Reviewer: Note updated code above...

>       (pseudobulk_frac_expressed + bulk_frac_expressed) > (params$expr_threshold * 2)

Reviewer: Since I was saying something about the intent, we might as well put that in the code more directly.
```suggestion
      mean(pseudobulk_frac_expressed, bulk_frac_expressed) > params$expr_threshold
```

# Make sure we have the same projects, in the same order
stopifnot(
  all.equal(names(pseudobulk_files), names(bulk_counts))
)

# also the tables with fraction expressed
bulk_frac_expr_df <- readr::read_tsv(
  file.path(data_dir, "fraction-expressed-bulk.tsv")
) |>
  dplyr::rename(bulk_frac_expressed = frac_samples_expressed)
  
named_projects <- purrr::set_names(pseudobulk_names, pseudobulk_names)
expressed_df <- named_projects |>
  purrr::map(
    \(project_id) {
      readr::read_tsv(
        file.path(data_dir, glue::glue("{project_id}_fraction-expressed-single-cell.tsv"))
      )
    }
  ) |>
  purrr::list_rbind(names_to = "project_id") |>
  dplyr::rename(pseudobulk_frac_expressed = frac_samples_expressed) |>
  dplyr::full_join(bulk_frac_expr_df, by = c("project_id", "ensembl_id"))
```

### Functions

This chunk defines a helper function to read and combine bulk and pseudobulk datasets for a single project.
```{r}
prepare_data <- function(pseudo_file, bulk_df, id, exclude_samples, expressed_df, filter = params$filter_genes) {
  
  # Find genes to keep if filtering is on
  if (filter) {
    keep_genes <- expressed_df |>
      dplyr::filter(
        project_id == id, 

>         pseudobulk_frac_expressed > 0,
>         bulk_frac_expressed > 0

Reviewer: Your current filter requires both bulk and pseudobulk to be expressed, which is not what we would want, as we could be quite interested in genes that are expressed in one but not the other. 
```suggestion
        pseudobulk_frac_expressed > 0 | bulk_frac_expressed > 0
```

That said, I don't think I agree with your decision to implement this as a boolean. If you instead made the filter numeric, you could do something as simple as:

```
pseudobulk_frac_expressed + bulk_frac_expressed > expr_threshold * 2
```

where setting having `expr_threshold = -1` would be equivalent to no filtering, and `expr_threshold = 0` would be the same as the above filter of expression in at least one sample. If we want to return with more filters, then we can do that later with no additional code changed in this script.

(Note that this also removes the need for the entire `if` block.)

Author: Sure, we can do this! The only thing that might get us is the interaction models may not run with a different threshold, but I say we cross that bridge if/when we ever get there?

Reviewer: I don't understand why it should have affected things at all. The filtering based on genes is across all samples, so the matrices are still complete.

Author: Frankly I didn't understand it myself but unfortunately don't immediately have the exact code under which it occurred. Maybe it was a different bug altogether then...
